rs/sns/testing/src/lib.rs

```diff
@@ -79,6 +84,79 @@ pub struct SnsProposalUpvoteArgs {
     pub wait: bool,
 }
 
+// A wrapper to parse 32-byte hex-encoded string.
+// Used to parse ICP account and subaccount.
+#[derive(Clone, Debug)]
+pub struct ParsedAccount(pub [u8; 32]);
+
+impl FromStr for ParsedAccount {
+    type Err = String;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        let hex_decoded = hex::decode(s).map_err(|e| e.to_string())?;
+        let bytes: [u8; 32] = hex_decoded
+            .try_into()
+            .map_err(|_| "Invalid length, must be 32 bytes")?;
+        Ok(Self(bytes))
+    }
+}
+
+#[derive(Debug, Subcommand)]
+pub enum TransferRecipientArg {
+    /// Transfer ICP to the specified account.
+    #[command(name = "--to")]
+    Account {
+        /// Recepient account, 32-byte hex-encoded.
+        account: ParsedAccount,
+    },
+    /// Transfer ICP to the specified principal's subaccount.
+    #[command(name = "--to-principal")]
+    Principal {
+        /// The recipient principal ID.
+        principal_id: PrincipalId,
+        /// The subaccount, 32-byte hex-encoded.
+        subaccount: Option<ParsedAccount>,
+    },
+}
+
+#[derive(Debug, Parser)]
+#[command(subcommand_value_name = "RECIPIENT")]
+pub struct TransferICPArgs {
+    /// The recipient of the transfer.
+    #[command(subcommand, name = "recipient")]
+    pub recipient: TransferRecipientArg,
+    /// The amount of e8s to transfer.
+    #[arg(long)]
+    pub amount: ParsedTokens,
+}
+
+#[derive(Clone, Debug)]
+pub struct ParsedTokens(pub Tokens);
+
+impl FromStr for ParsedTokens {
+    type Err = String;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        let decimal_tokens = Decimal::from_str(s).map_err(|e| e.to_string())?;
+        if decimal_tokens.scale() > 8 {
+            return Err(
+                "The number of decimal places for ICP token must be less than or equal to 8"
+                    .to_string(),
+            );
+        }
+        if decimal_tokens.is_sign_negative() {
+            return Err("The amount of ICP tokens must be positive".to_string());
+        }
+        let e8s: i128 = decimal_tokens.mantissa() * 10i128.pow(8 - decimal_tokens.scale());
```

Comment:
Can you please describe what is being done here? May be also include it as a comment? Also what do you think about adding some simple tests for this code in this module itself?

Reply:
Added a comment and a bunch of unit tests for `FromStr` implementation